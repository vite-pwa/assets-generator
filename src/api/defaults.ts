import type { PngOptions, ResizeOptions } from 'sharp'
import type { AssetType } from '../types.ts'

export const defaultResizeOptions = {
  fit: 'contain',
  background: 'white',
} satisfies ResizeOptions

export const defaultDarkResizeOptions = {
  fit: 'contain',
  background: 'black',
} satisfies ResizeOptions

export const defaultPngCompressionOptions: PngOptions = {
  compressionLevel: 9,
  quality: 60,
}

export const defaultPngOptions: Record<AssetType, { padding: number, resizeOptions: ResizeOptions }> = {
  transparent: { padding: 0.05, resizeOptions: { fit: 'contain', background: 'transparent' } },
  maskable: { padding: 0.3, resizeOptions: { fit: 'contain', background: 'white' } },
  apple: { padding: 0.3, resizeOptions: { fit: 'contain', background: 'white' } },
}

export function createResizeOptions(dark: boolean, options: ResizeOptions) {
  return {
    ...(dark ? defaultDarkResizeOptions : defaultResizeOptions),
    ...options,
  } satisfies ResizeOptions
}

export function createPngCompressionOptions(options: PngOptions) {
  return {
    ...defaultPngCompressionOptions,
    ...options,
  } satisfies PngOptions
}
