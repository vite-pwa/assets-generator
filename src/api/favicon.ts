import type { GenerateFaviconOptionsType, GenerateFaviconType, ImageSourceInput } from './types.ts'
import sharp from 'sharp'
import { encode } from 'sharp-ico'

export async function generateFavicon<Format extends GenerateFaviconType>(
  format: Format,
  image: ImageSourceInput,
  options?: GenerateFaviconOptionsType<Format>,
) {
  return encode([await sharp(image).toFormat(format, options).toBuffer()])
}

if (import.meta.vitest) {
  const { expect, it } = import.meta.vitest
  it('should generate a favicon asset', async () => {
    const result = await generateFavicon(
      'png',
      'playground/pwa/public/favicon.svg',
    )
    expect(result).toBeDefined()
  })
}
