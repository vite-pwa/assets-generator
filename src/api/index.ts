export * from './types.ts'
// eslint-disable-next-line perfectionist/sort-exports
export * from './defaults.ts'
export * from './utils.ts'
// eslint-disable-next-line perfectionist/sort-exports
export * from './html.ts'
export * from './maskable.ts'
export * from './transparent.ts'
// eslint-disable-next-line perfectionist/sort-exports
export * from './favicon.ts'
