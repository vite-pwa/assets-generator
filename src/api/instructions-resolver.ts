import type {
  BuiltInPreset,
  Preset,
  ResolvedAssets,
} from '../config.ts'
import {
  defaultAssetName,
  defaultPngCompressionOptions,
} from '../config.ts'
import { toResolvedAsset } from '../utils.ts'
import type { HtmlLinkPreset } from './html.ts'
import {
  createFaviconHtmlLink,
} from './html.ts'
import type { ImageAssets, ImageAssetsInstructions } from './types.ts'
import { resolveAppleSplashScreensInstructions } from './apple-icons-helper.ts'
import { resolveMaskableIcons, resolveTransparentIcons } from './icons-resolver-helper.ts'

export async function resolveInstructions(imageAssets: ImageAssets) {
  const {
    imageResolver,
    imageName,
    originalName,
    preset = 'minimal',
    faviconPreset,
  } = imageAssets

  const [usePreset, htmlPreset] = await resolvePreset(preset, faviconPreset)

  const {
    assetName = defaultAssetName,
    png = defaultPngCompressionOptions,
    appleSplashScreens: useAppleSplashScreens,
  } = usePreset

  const assets: ResolvedAssets = {
    assets: {
      transparent: toResolvedAsset('transparent', usePreset.transparent),
      maskable: toResolvedAsset('maskable', usePreset.maskable),
      apple: toResolvedAsset('apple', usePreset.apple),
    },
    png,
    assetName,
  }

  const instructions = {
    image: imageName,
    originalName,
    favicon: {},
    transparent: {},
    maskable: {},
    apple: {},
    appleSplashScreen: {},
  } as ImageAssetsInstructions

  const image = await imageResolver()

  resolveTransparentIcons(imageAssets, image, assets, htmlPreset, instructions)
  resolveMaskableIcons('maskable', imageAssets, image, assets, htmlPreset, instructions)
  resolveMaskableIcons('apple', imageAssets, image, assets, htmlPreset, instructions)

  if (imageName.endsWith('.svg')) {
    const name = imageAssets.resolveSvgName(imageName)
    const url = `${imageAssets.basePath}${name}`
    instructions.favicon[url] = {
      name,
      url,
      width: 0,
      height: 0,
      mimeType: 'image/svg+xml',
      link: createFaviconHtmlLink('string', htmlPreset, {
        name,
        basePath: imageAssets.basePath,
        xhtml: imageAssets.htmlLinks.xhtml,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      linkObject: createFaviconHtmlLink('link', htmlPreset, {
        name,
        basePath: imageAssets.basePath,
        xhtml: imageAssets.htmlLinks.xhtml,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      buffer: () => Promise.resolve(image),
    }
  }

  resolveAppleSplashScreensInstructions(imageAssets, instructions, useAppleSplashScreens)

  return instructions
}

async function resolvePreset(preset: BuiltInPreset | Preset, faviconPreset?: HtmlLinkPreset): Promise<[Preset, HtmlLinkPreset]> {
  if (typeof preset === 'object')
    return [preset, faviconPreset ?? 'default']

  switch (preset) {
    case 'minimal':
      return [await import('../presets/minimal.ts').then(m => m.minimalPreset), 'default']
    case 'minimal-2023':
      return [await import('../presets/minimal-2023.ts').then(m => m.minimal2023Preset), '2023']
    default:
      throw new Error(`Preset ${preset} not yet implemented`)
  }
}
