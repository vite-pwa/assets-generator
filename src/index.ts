export * from './types'
export * from './utils.ts'
// eslint-disable-next-line perfectionist/sort-exports
export * from './api/defaults.ts'
