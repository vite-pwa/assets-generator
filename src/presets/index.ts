export * from './minimal.ts'
export * from './minimal-2023.ts'
