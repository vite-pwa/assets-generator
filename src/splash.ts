import type { PngOptions, ResizeOptions } from 'sharp'
import type { AppleDeviceSize, AppleSplashScreenName, AppleSplashScreens, AppleTouchStartupImageOptions } from './types.ts'
import type { Preset } from './preset.ts'

export type AppleDeviceName =
    | 'iPad Pro 12.9"'
    | 'iPad Pro 11"'
    | 'iPad Pro 10.5"'
    | 'iPad Pro 9.7"'
    | 'iPad mini 7.9"'
    | 'iPad Air 10.5"'
    | 'iPad Air 9.7"'
    | 'iPad 10.2"'
    | 'iPad 9.7"'
    | 'iPhone 14 Pro Max'
    | 'iPhone 14 Pro'
    | 'iPhone 14 Plus'
    | 'iPhone 14'
    | 'iPhone 13 Pro Max'
    | 'iPhone 13 Pro'
    | 'iPhone 13'
    | 'iPhone 13 mini'
    | 'iPhone 12 Pro Max'
    | 'iPhone 12 Pro'
    | 'iPhone 12'
    | 'iPhone 12 mini'
    | 'iPhone 11 Pro Max'
    | 'iPhone 11 Pro'
    | 'iPhone 11'
    | 'iPhone XS Max'
    | 'iPhone XS'
    | 'iPhone XR'
    | 'iPhone X'
    | 'iPhone 8 Plus'
    | 'iPhone 8'
    | 'iPhone 7 Plus'
    | 'iPhone 7'
    | 'iPhone 6s Plus'
    | 'iPhone 6s'
    | 'iPhone 6 Plus'
    | 'iPhone 6'
    | 'iPhone SE 4.7"'
    | 'iPhone SE 4"'
    | 'iPod touch 5th generation and later'

export interface AppleSplashScreen extends Record<AppleDeviceName, AppleDeviceSize> {}

export const appleSplashScreenSizes: AppleSplashScreen = {
  'iPad Pro 12.9"': { width: 2048, height: 2732, scaleFactor: 2 },
  'iPad Pro 11"': { width: 1668, height: 2388, scaleFactor: 2 },
  'iPad Pro 10.5"': { width: 1668, height: 2388, scaleFactor: 2 },
  'iPad Pro 9.7"': { width: 1536, height: 2048, scaleFactor: 2 },
  'iPad mini 7.9"': { width: 1536, height: 2048, scaleFactor: 2 },
  'iPad Air 10.5"': { width: 1668, height: 2224, scaleFactor: 2 },
  'iPad Air 9.7"': { width: 1536, height: 2048, scaleFactor: 2 },
  'iPad 10.2"': { width: 1620, height: 2160, scaleFactor: 2 },
  'iPad 9.7"': { width: 1536, height: 2048, scaleFactor: 2 },
  'iPhone 14 Pro Max': { width: 1290, height: 2796, scaleFactor: 3 },
  'iPhone 14 Pro': { width: 1179, height: 2556, scaleFactor: 3 },
  'iPhone 14 Plus': { width: 1284, height: 2778, scaleFactor: 3 },
  'iPhone 14': { width: 1170, height: 2532, scaleFactor: 3 },
  'iPhone 13 Pro Max': { width: 1284, height: 2778, scaleFactor: 3 },
  'iPhone 13 Pro': { width: 1170, height: 2532, scaleFactor: 3 },
  'iPhone 13': { width: 1170, height: 2532, scaleFactor: 3 },
  'iPhone 13 mini': { width: 1125, height: 2436, scaleFactor: 3 },
  'iPhone 12 Pro Max': { width: 1284, height: 2778, scaleFactor: 3 },
  'iPhone 12 Pro': { width: 1170, height: 2532, scaleFactor: 3 },
  'iPhone 12': { width: 1170, height: 2532, scaleFactor: 3 },
  'iPhone 12 mini': { width: 1125, height: 2436, scaleFactor: 3 },
  'iPhone 11 Pro Max': { width: 1242, height: 2688, scaleFactor: 3 },
  'iPhone 11 Pro': { width: 1125, height: 2436, scaleFactor: 3 },
  'iPhone 11': { width: 828, height: 1792, scaleFactor: 2 },
  'iPhone XS Max': { width: 1242, height: 2688, scaleFactor: 3 },
  'iPhone XS': { width: 1125, height: 2436, scaleFactor: 3 },
  'iPhone XR': { width: 828, height: 1792, scaleFactor: 2 },
  'iPhone X': { width: 1125, height: 2436, scaleFactor: 3 },
  'iPhone 8 Plus': { width: 1242, height: 2208, scaleFactor: 3 },
  'iPhone 8': { width: 750, height: 1334, scaleFactor: 2 },
  'iPhone 7 Plus': { width: 1242, height: 2208, scaleFactor: 3 },
  'iPhone 7': { width: 750, height: 1334, scaleFactor: 2 },
  'iPhone 6s Plus': { width: 1242, height: 2208, scaleFactor: 3 },
  'iPhone 6s': { width: 750, height: 1334, scaleFactor: 2 },
  'iPhone 6 Plus': { width: 1242, height: 2208, scaleFactor: 3 },
  'iPhone 6': { width: 750, height: 1334, scaleFactor: 2 },
  'iPhone SE 4.7"': { width: 750, height: 1334, scaleFactor: 2 },
  'iPhone SE 4"': { width: 640, height: 1136, scaleFactor: 2 },
  'iPod touch 5th generation and later': { width: 640, height: 1136, scaleFactor: 2 },
}

export function defaultSplashScreenName(landscape: boolean, size: AppleDeviceSize, dark?: boolean) {
  return `apple-splash-${landscape ? 'landscape' : 'portrait'}-${typeof dark === 'boolean' ? (dark ? 'dark-' : 'light-') : ''}${size.width}x${size.height}.png`
}

export const AllAppleDeviceNames = Array.from(Object.keys(appleSplashScreenSizes).map(k => k as AppleDeviceName))

export function createAppleSplashScreens(
  options: {
    padding?: number
    resizeOptions?: ResizeOptions
    darkResizeOptions?: ResizeOptions
    linkMediaOptions?: AppleTouchStartupImageOptions
    png?: PngOptions
    name?: AppleSplashScreenName
  } = {},
  devices: AppleDeviceName[] = AllAppleDeviceNames,
) {
  const {
    padding,
    resizeOptions,
    darkResizeOptions,
    linkMediaOptions,
    png,
    name,
  } = options

  return {
    sizes: devices.map(deviceName => appleSplashScreenSizes[deviceName]),
    padding,
    resizeOptions,
    darkResizeOptions,
    linkMediaOptions,
    png,
    name,
  } satisfies AppleSplashScreens
}

export function combinePresetAndAppleSplashScreens(
  preset: Preset,
  options: {
    padding?: number
    resizeOptions?: ResizeOptions
    darkResizeOptions?: ResizeOptions
    linkMediaOptions?: AppleTouchStartupImageOptions
    png?: PngOptions
    name?: AppleSplashScreenName
  } = {},
  devices: AppleDeviceName[] = AllAppleDeviceNames,
) {
  return {
    ...preset,
    appleSplashScreens: createAppleSplashScreens(options, devices),
  } satisfies Preset
}
