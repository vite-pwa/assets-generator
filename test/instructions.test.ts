import { readFile } from 'node:fs/promises'
import { basename } from 'node:path'
import { describe, expect, it } from 'vitest'
import { minimal2023Preset } from '../src/presets'
import { createAppleSplashScreens } from '../src/splash'
import { resolveInstructions } from '../src/api/instructions-resolver'

describe('instructions', () => {
  it ('resolve instructions', async () => {
    const instructions = await resolveInstructions({
      imageResolver: () => readFile('playground/pwa/public/favicon.svg'),
      imageName: 'playground/pwa/public/favicon.svg',
      preset: 'minimal-2023',
      htmlLinks: {
        xhtml: false,
        includeId: false,
      },
      basePath: '/',
      resolveSvgName: name => basename(name),
    })
    expect(instructions).toMatchSnapshot()
  })
  it ('resolve instructions with apple splash screen icons', async () => {
    const instructions = await resolveInstructions({
      imageResolver: () => readFile('playground/pwa/public/favicon.svg'),
      imageName: 'playground/pwa/public/favicon.svg',
      faviconPreset: '2023',
      preset: {
        ...minimal2023Preset,
        appleSplashScreens: createAppleSplashScreens({
          padding: 0.3,
          resizeOptions: { fit: 'contain', background: 'white' },
          darkResizeOptions: { fit: 'contain', background: 'black' },
          linkMediaOptions: {
            log: true,
            addMediaScreen: true,
            xhtml: true,
          },
        }, ['iPad Air 9.7"']),
      },
      htmlLinks: {
        xhtml: false,
        includeId: false,
      },
      basePath: '/',
      resolveSvgName: name => basename(name),
    })
    expect(instructions).toMatchSnapshot()
  })
  it ('resolve instructions with apple splash screen icons with custom base url', async () => {
    const instructions = await resolveInstructions({
      imageResolver: () => readFile('playground/pwa/public/favicon.svg'),
      imageName: 'playground/pwa/public/favicon.svg',
      faviconPreset: '2023',
      basePath: '/test/',
      preset: {
        ...minimal2023Preset,
        appleSplashScreens: createAppleSplashScreens({
          padding: 0.3,
          resizeOptions: { fit: 'contain', background: 'white' },
          darkResizeOptions: { fit: 'contain', background: 'black' },
          linkMediaOptions: {
            log: true,
            addMediaScreen: true,
            xhtml: true,
          },
        }, ['iPad Air 9.7"']),
      },
      htmlLinks: {
        xhtml: false,
        includeId: false,
      },
      resolveSvgName: name => basename(name),
    })
    expect(instructions).toMatchSnapshot()
  })
})
