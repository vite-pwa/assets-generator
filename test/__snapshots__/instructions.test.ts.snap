// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`instructions > resolve instructions 1`] = `
{
  "apple": {
    "/apple-touch-icon-180x180.png": {
      "buffer": [Function],
      "height": 180,
      "link": "<link rel="apple-touch-icon" href="/apple-touch-icon-180x180.png">",
      "linkObject": {
        "href": "/apple-touch-icon-180x180.png",
        "id": "ati-180-180",
        "rel": "apple-touch-icon",
      },
      "mimeType": "image/png",
      "name": "apple-touch-icon-180x180.png",
      "url": "/apple-touch-icon-180x180.png",
      "width": 180,
    },
  },
  "appleSplashScreen": {},
  "favicon": {
    "/favicon.ico": {
      "buffer": [Function],
      "height": 48,
      "link": "<link rel="icon" href="/favicon.ico" sizes="48x48">",
      "linkObject": {
        "href": "/favicon.ico",
        "id": "fav-48x48",
        "rel": "icon",
        "sizes": "48x48",
      },
      "mimeType": "image/x-icon",
      "name": "favicon.ico",
      "url": "/favicon.ico",
      "width": 48,
    },
    "/favicon.svg": {
      "buffer": [Function],
      "height": 0,
      "link": "<link rel="icon" href="/favicon.svg" sizes="any" type="image/svg+xml">",
      "linkObject": {
        "href": "/favicon.svg",
        "id": "fav-svg",
        "rel": "icon",
        "sizes": "any",
        "type": "image/svg+xml",
      },
      "mimeType": "image/svg+xml",
      "name": "favicon.svg",
      "url": "/favicon.svg",
      "width": 0,
    },
  },
  "image": "playground/pwa/public/favicon.svg",
  "maskable": {
    "/maskable-icon-512x512.png": {
      "buffer": [Function],
      "height": 512,
      "mimeType": "image/png",
      "name": "maskable-icon-512x512.png",
      "url": "/maskable-icon-512x512.png",
      "width": 512,
    },
  },
  "originalName": undefined,
  "transparent": {
    "/pwa-192x192.png": {
      "buffer": [Function],
      "height": 192,
      "mimeType": "image/png",
      "name": "pwa-192x192.png",
      "url": "/pwa-192x192.png",
      "width": 192,
    },
    "/pwa-512x512.png": {
      "buffer": [Function],
      "height": 512,
      "mimeType": "image/png",
      "name": "pwa-512x512.png",
      "url": "/pwa-512x512.png",
      "width": 512,
    },
    "/pwa-64x64.png": {
      "buffer": [Function],
      "height": 64,
      "mimeType": "image/png",
      "name": "pwa-64x64.png",
      "url": "/pwa-64x64.png",
      "width": 64,
    },
  },
}
`;

exports[`instructions > resolve instructions with apple splash screen icons 1`] = `
{
  "apple": {
    "/apple-touch-icon-180x180.png": {
      "buffer": [Function],
      "height": 180,
      "link": "<link rel="apple-touch-icon" href="/apple-touch-icon-180x180.png">",
      "linkObject": {
        "href": "/apple-touch-icon-180x180.png",
        "id": "ati-180-180",
        "rel": "apple-touch-icon",
      },
      "mimeType": "image/png",
      "name": "apple-touch-icon-180x180.png",
      "url": "/apple-touch-icon-180x180.png",
      "width": 180,
    },
  },
  "appleSplashScreen": {
    "/apple-splash-landscape-dark-2048x1536.png": {
      "buffer": [Function],
      "height": 1536,
      "link": "<link rel="apple-touch-startup-image" media="screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="/apple-splash-landscape-dark-2048x1536.png" />",
      "linkObject": {
        "href": "/apple-splash-landscape-dark-2048x1536.png",
        "id": "atsi-1536-2048-2-dark",
        "media": "screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-landscape-dark-2048x1536.png",
      "url": "/apple-splash-landscape-dark-2048x1536.png",
      "width": 2048,
    },
    "/apple-splash-landscape-light-2048x1536.png": {
      "buffer": [Function],
      "height": 1536,
      "link": "<link rel="apple-touch-startup-image" media="screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="/apple-splash-landscape-light-2048x1536.png" />",
      "linkObject": {
        "href": "/apple-splash-landscape-light-2048x1536.png",
        "id": "atsi-1536-2048-2-light",
        "media": "screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-landscape-light-2048x1536.png",
      "url": "/apple-splash-landscape-light-2048x1536.png",
      "width": 2048,
    },
    "/apple-splash-portrait-dark-1536x2048.png": {
      "buffer": [Function],
      "height": 2048,
      "link": "<link rel="apple-touch-startup-image" media="screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="/apple-splash-portrait-dark-1536x2048.png" />",
      "linkObject": {
        "href": "/apple-splash-portrait-dark-1536x2048.png",
        "id": "atsi-1536-2048-2-dark",
        "media": "screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-portrait-dark-1536x2048.png",
      "url": "/apple-splash-portrait-dark-1536x2048.png",
      "width": 1536,
    },
    "/apple-splash-portrait-light-1536x2048.png": {
      "buffer": [Function],
      "height": 2048,
      "link": "<link rel="apple-touch-startup-image" media="screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="/apple-splash-portrait-light-1536x2048.png" />",
      "linkObject": {
        "href": "/apple-splash-portrait-light-1536x2048.png",
        "id": "atsi-1536-2048-2-light",
        "media": "screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-portrait-light-1536x2048.png",
      "url": "/apple-splash-portrait-light-1536x2048.png",
      "width": 1536,
    },
  },
  "favicon": {
    "/favicon.ico": {
      "buffer": [Function],
      "height": 48,
      "link": "<link rel="icon" href="/favicon.ico" sizes="48x48">",
      "linkObject": {
        "href": "/favicon.ico",
        "id": "fav-48x48",
        "rel": "icon",
        "sizes": "48x48",
      },
      "mimeType": "image/x-icon",
      "name": "favicon.ico",
      "url": "/favicon.ico",
      "width": 48,
    },
    "/favicon.svg": {
      "buffer": [Function],
      "height": 0,
      "link": "<link rel="icon" href="/favicon.svg" sizes="any" type="image/svg+xml">",
      "linkObject": {
        "href": "/favicon.svg",
        "id": "fav-svg",
        "rel": "icon",
        "sizes": "any",
        "type": "image/svg+xml",
      },
      "mimeType": "image/svg+xml",
      "name": "favicon.svg",
      "url": "/favicon.svg",
      "width": 0,
    },
  },
  "image": "playground/pwa/public/favicon.svg",
  "maskable": {
    "/maskable-icon-512x512.png": {
      "buffer": [Function],
      "height": 512,
      "mimeType": "image/png",
      "name": "maskable-icon-512x512.png",
      "url": "/maskable-icon-512x512.png",
      "width": 512,
    },
  },
  "originalName": undefined,
  "transparent": {
    "/pwa-192x192.png": {
      "buffer": [Function],
      "height": 192,
      "mimeType": "image/png",
      "name": "pwa-192x192.png",
      "url": "/pwa-192x192.png",
      "width": 192,
    },
    "/pwa-512x512.png": {
      "buffer": [Function],
      "height": 512,
      "mimeType": "image/png",
      "name": "pwa-512x512.png",
      "url": "/pwa-512x512.png",
      "width": 512,
    },
    "/pwa-64x64.png": {
      "buffer": [Function],
      "height": 64,
      "mimeType": "image/png",
      "name": "pwa-64x64.png",
      "url": "/pwa-64x64.png",
      "width": 64,
    },
  },
}
`;

exports[`instructions > resolve instructions with apple splash screen icons with custom base url 1`] = `
{
  "apple": {
    "/test/apple-touch-icon-180x180.png": {
      "buffer": [Function],
      "height": 180,
      "link": "<link rel="apple-touch-icon" href="/test/apple-touch-icon-180x180.png">",
      "linkObject": {
        "href": "/test/apple-touch-icon-180x180.png",
        "id": "ati-180-180",
        "rel": "apple-touch-icon",
      },
      "mimeType": "image/png",
      "name": "apple-touch-icon-180x180.png",
      "url": "/test/apple-touch-icon-180x180.png",
      "width": 180,
    },
  },
  "appleSplashScreen": {
    "/test/apple-splash-landscape-dark-2048x1536.png": {
      "buffer": [Function],
      "height": 1536,
      "link": "<link rel="apple-touch-startup-image" media="screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="/test/apple-splash-landscape-dark-2048x1536.png" />",
      "linkObject": {
        "href": "/test/apple-splash-landscape-dark-2048x1536.png",
        "id": "atsi-1536-2048-2-dark",
        "media": "screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-landscape-dark-2048x1536.png",
      "url": "/test/apple-splash-landscape-dark-2048x1536.png",
      "width": 2048,
    },
    "/test/apple-splash-landscape-light-2048x1536.png": {
      "buffer": [Function],
      "height": 1536,
      "link": "<link rel="apple-touch-startup-image" media="screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="/test/apple-splash-landscape-light-2048x1536.png" />",
      "linkObject": {
        "href": "/test/apple-splash-landscape-light-2048x1536.png",
        "id": "atsi-1536-2048-2-light",
        "media": "screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-landscape-light-2048x1536.png",
      "url": "/test/apple-splash-landscape-light-2048x1536.png",
      "width": 2048,
    },
    "/test/apple-splash-portrait-dark-1536x2048.png": {
      "buffer": [Function],
      "height": 2048,
      "link": "<link rel="apple-touch-startup-image" media="screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="/test/apple-splash-portrait-dark-1536x2048.png" />",
      "linkObject": {
        "href": "/test/apple-splash-portrait-dark-1536x2048.png",
        "id": "atsi-1536-2048-2-dark",
        "media": "screen and (prefers-color-scheme: dark) and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-portrait-dark-1536x2048.png",
      "url": "/test/apple-splash-portrait-dark-1536x2048.png",
      "width": 1536,
    },
    "/test/apple-splash-portrait-light-1536x2048.png": {
      "buffer": [Function],
      "height": 2048,
      "link": "<link rel="apple-touch-startup-image" media="screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="/test/apple-splash-portrait-light-1536x2048.png" />",
      "linkObject": {
        "href": "/test/apple-splash-portrait-light-1536x2048.png",
        "id": "atsi-1536-2048-2-light",
        "media": "screen and (device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)",
        "rel": "apple-touch-startup-image",
      },
      "mimeType": "image/png",
      "name": "apple-splash-portrait-light-1536x2048.png",
      "url": "/test/apple-splash-portrait-light-1536x2048.png",
      "width": 1536,
    },
  },
  "favicon": {
    "/test/favicon.ico": {
      "buffer": [Function],
      "height": 48,
      "link": "<link rel="icon" href="/test/favicon.ico" sizes="48x48">",
      "linkObject": {
        "href": "/test/favicon.ico",
        "id": "fav-48x48",
        "rel": "icon",
        "sizes": "48x48",
      },
      "mimeType": "image/x-icon",
      "name": "favicon.ico",
      "url": "/test/favicon.ico",
      "width": 48,
    },
    "/test/favicon.svg": {
      "buffer": [Function],
      "height": 0,
      "link": "<link rel="icon" href="/test/favicon.svg" sizes="any" type="image/svg+xml">",
      "linkObject": {
        "href": "/test/favicon.svg",
        "id": "fav-svg",
        "rel": "icon",
        "sizes": "any",
        "type": "image/svg+xml",
      },
      "mimeType": "image/svg+xml",
      "name": "favicon.svg",
      "url": "/test/favicon.svg",
      "width": 0,
    },
  },
  "image": "playground/pwa/public/favicon.svg",
  "maskable": {
    "/test/maskable-icon-512x512.png": {
      "buffer": [Function],
      "height": 512,
      "mimeType": "image/png",
      "name": "maskable-icon-512x512.png",
      "url": "/test/maskable-icon-512x512.png",
      "width": 512,
    },
  },
  "originalName": undefined,
  "transparent": {
    "/test/pwa-192x192.png": {
      "buffer": [Function],
      "height": 192,
      "mimeType": "image/png",
      "name": "pwa-192x192.png",
      "url": "/test/pwa-192x192.png",
      "width": 192,
    },
    "/test/pwa-512x512.png": {
      "buffer": [Function],
      "height": 512,
      "mimeType": "image/png",
      "name": "pwa-512x512.png",
      "url": "/test/pwa-512x512.png",
      "width": 512,
    },
    "/test/pwa-64x64.png": {
      "buffer": [Function],
      "height": 64,
      "mimeType": "image/png",
      "name": "pwa-64x64.png",
      "url": "/test/pwa-64x64.png",
      "width": 64,
    },
  },
}
`;
